### PR TITLE
feat(api): Add automatic quota throttling

### PR DESCRIPTION
**Pull request by rifont**
### What changed? Why was the change needed?
* Integrate the quota limiting PR from EE repo
* Add feature flag for quota limiting
* Rename feature flag for event limiting to align with other naming and purpose
* Refactor cache key builders to use new folder hierarchy to separate builder pattern from use-cases
* Add new cache key builders for quota limit cache

### Screenshots
_Passing key builder Tests_
image

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Related enterprise PR

https://github.com/novuhq/packages-enterprise/pull/133

### Special notes for your reviewer

</details>

### CODE REVIEW

File: apps/web/src/components/nav/SettingsNavMenu.tsx [hunk 1]
```diff
@@ -71,7 +71,7 @@ export const SettingsNavMenu: React.FC = () => {
           testId="side-nav-settings-branding-link"
         ></NavMenuLinkButton>
         <NavMenuLinkButton
-          label="Billing plans"
+          label="Billing"
```
Comment (rifont) [inline apps/web/src/components/nav/SettingsNavMenu.tsx:74]: This was a regression during IA feature development.

File: apps/api/src/app/resource-limiting/guards/resource-throttler.guard.ts [hunk 2]
```diff
@@ -47,7 +48,7 @@ export class ResourceThrottlerInterceptor implements NestInterceptor {
             organizationId,
             environmentId: 'system',
             userId: 'system',
-            key: FeatureFlagsKeysEnum.IS_EVENT_RESOURCE_LIMITING_ENABLED,
+            key: FeatureFlagsKeysEnum.IS_EVENT_QUOTA_LIMITING_ENABLED,
```
Comment (rifont) [inline apps/api/src/app/resource-limiting/guards/resource-throttler.guard.ts:51]: Renaming to align with intended purpose, which is to limit resource quotas.

File: packages/application-generic/src/services/cache/key-builders/crypto.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { createHash as createHashCrypto } from 'crypto';
+
+export const createHash = (str: string): string => {
+  const hash = createHashCrypto('sha256');
+  hash.update(str);
+
+  return hash.digest('hex');
+};
```
Comment (rifont) [inline packages/application-generic/src/services/cache/key-builders/crypto.ts:8]: Separating into a new file to keep independent from the key builder use-cases.

File: packages/application-generic/src/services/cache/key-builders/entities.ts [hunk 4]
```diff
@@ -1,173 +1,176 @@
...
-  buildNotificationTemplateIdentifierKey,
-  buildEnvironmentByApiKey,
-  buildKeyById,
-  buildGroupedBlueprintsKey,
-  buildAuthServiceKey,
-  buildMaximumApiRateLimitKey,
-  buildEvaluateApiRateLimitKey,
-  buildServiceConfigApiRateLimitMaximumKey,
-  buildVariablesKey,
-};
```
Comment (rifont) [inline packages/application-generic/src/services/cache/key-builders/entities.ts:173]: It's a bit of a code smell to update here, so we instead export each of the required builders in-line.

File: packages/application-generic/src/services/cache/key-builders/queries.ts
Comment (rifont) [inline packages/application-generic/src/services/cache/key-builders/queries.ts:1]: This file only has internal API changes, no change to external exports.

File: packages/application-generic/src/services/cache/key-builders/entities.ts
Comment (rifont) [inline packages/application-generic/src/services/cache/key-builders/entities.ts:1]: This file only has internal API changes, no changes to external exports.

File: packages/application-generic/src/services/cache/key-builders/shared.ts
Comment (rifont) [inline packages/application-generic/src/services/cache/key-builders/shared.ts:1]: This file has comments added to provide a description of when to modify and use each base key builder. There are changes to the exports of these builders which are in-turn, consumed by the `entities` and `queries`.

File: packages/application-generic/src/services/cache/key-builders/shared.ts [hunk 5]
```diff
@@ -87,15 +241,26 @@ export enum IdentifierPrefixEnum {
   GROUPED_BLUEPRINT = 'g_b',
   API_RATE_LIMIT_CATEGORY = 'a_r_l_c',
   SERVICE_CONFIG = 's_c',
+  RESOURCE_TYPE = 'r_t',
```
Comment (rifont) [inline packages/application-generic/src/services/cache/key-builders/shared.ts:244]: New prefix to handle storage of quota usage for each resource.

File: packages/application-generic/src/services/cache/key-builders/identifiers.ts
Comment (rifont) [inline packages/application-generic/src/services/cache/key-builders/identifiers.ts:1]: Refactoring all the identifiers out into a separate file.

Comment (djabarovgeorge) [reply packages/application-generic/src/services/cache/key-builders/identifiers.ts:1]: Love it!

File: packages/application-generic/src/services/cache/key-builders/shared.ts [hunk 6]
```diff
@@ -1,45 +1,79 @@
...
+  CacheKeyPrefixEnum,
+  CacheKeyTypeEnum,
+  IdentifierPrefixEnum,
+  OrgScopePrefixEnum,
+  ServiceConfigIdentifierEnum,
+} from './identifiers';
+
+/**
+ * Wraps the entire prefix string with curly braces. This has the effect of ensuring
+ * that the entire prefix string is treated as a single key part by Redis.
```
Comment (djabarovgeorge) [inline packages/application-generic/src/services/cache/key-builders/shared.ts:17]: Love this one!

File: packages/application-generic/src/services/cache/key-builders/shared.ts [hunk 7]
```diff
@@ -1,45 +1,79 @@
 /**
- * Use this to build a key for entities that are scoped to an environment
+ ***************************************
+ *         BASE KEY BUILDERS
+ ***************************************
+ */
```
Comment (djabarovgeorge) [inline packages/application-generic/src/services/cache/key-builders/shared.ts:5]: Maybe we could create a separate file for every section that has such a title, wdyt?

**Review by djabarovgeorge — APPROVED** [review]
LGTM, left a couple of comments for your review.
all of them as `thought`s

**Pull request merged**